### PR TITLE
Updates to fix coercion errors. DSDEEPB-727

### PR DESCRIPTION
**Pull request by kshakir**
Fixed Symbol doubly qualifying the symbol names. (thx mcovarr/scottfrazer)
Added `WdlType.fromRawString`, with test against respective `WdlValue.toRawString`.
`DummyDataAccess` replaced with using `DataAccess` instances, with cleanup of connections.
When creating in memory databases will create unique `DataAccess` instances, just like Dummy.
TestSlickDatabase now prints a warning, instead of an error, when unable to connect to MySql.

### CODE REVIEW

File: src/test/scala/cromwell/engine/WorkflowManagerActorSpec.scala [hunk 1]
```diff
@@ -39,42 +42,43 @@ class WorkflowManagerActorSpec extends CromwellTestkitSpec("WorkflowManagerActor
 
       val actual = outputs.map { case (k, WdlString(string)) => k -> string }
       actual shouldEqual Map(HelloWorld.OutputKey -> HelloWorld.OutputValue)
+      Await.result(dataAccess.shutdown(), Duration.Inf)
```
Comment (mcovarr) [inline src/test/scala/cromwell/engine/WorkflowManagerActorSpec.scala]: could this use a loan pattern for the `DataAccess`?  (Warning: did not test this)

```
with { dataAccess =>
  doStuff()
}

def with[T](work: DataAccess => T): T = {
  val dataAccess = DataAccess()
  work(dataAccess)
  Await.result(dataAccess.shutdown(), Duration.Inf)
}
```

File: src/test/scala/cromwell/engine/WorkflowManagerActorSpec.scala [hunk 2]
```diff
@@ -39,42 +42,43 @@ class WorkflowManagerActorSpec extends CromwellTestkitSpec("WorkflowManagerActor
...
-          Future.successful {
-            workflows.map { w =>
-              WorkflowInfo(w.workflowId, w.wdlSource, w.jsonInputs)
-            }
-          }
-        }
-      }
+      val dataAccess = DataAccess()
+      import ExecutionContext.Implicits.global
+      val setupFuture = Future.sequence(workflows map { kv =>
```
Comment (mcovarr) [inline src/test/scala/cromwell/engine/WorkflowManagerActorSpec.scala]: ```case (workflowId, workflowState) => ```?

File: src/test/scala/cromwell/engine/db/DummyDataAccess.scala [hunk 3]
```diff
@@ -1,99 +0,0 @@
-package cromwell.engine.db
```
Comment (mcovarr) [inline src/test/scala/cromwell/engine/db/DummyDataAccess.scala]: good riddance :smile:

File: src/main/scala/cromwell/engine/db/QueryWorkflowExecutionResult.scala [hunk 4]
```diff
@@ -1,11 +0,0 @@
-package cromwell.engine.db
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/db/QueryWorkflowExecutionResult.scala]: oops hadn't noticed this was still around...

File: src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala [hunk 5]
```diff
@@ -36,6 +35,24 @@ object SlickDataAccess {
   implicit class StringToClob(val str: String) extends AnyVal {
     def toClob: Clob = new SerialClob(str.toCharArray)
   }
+
+  implicit class ConfigWithUniqueSchema(val config: Config) extends AnyVal {
+    def withUniqueSchema: Config = {
+      val url = config.getString("url")
+      if (url.contains("${slick.uniqueSchema}")) {
+        // Config wasn't updating with a simple withValue/withFallback.
+        // So instead, do a bit of extra work to insert the generated schema name in the url.
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: I wish I knew Typesafe Config well enough to suggest a workaround... :cry:

File: src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala [hunk 6]
```diff
@@ -30,15 +30,15 @@ class TestSlickDatabase(configPath: String) {
...
-          false
-      } andThen {
+      database.run(SimpleDBIO(_.connection.isValid(1))) andThen {
         case _ =>
           log.debug("Closing test connection setup for " + configPath)
           database.close()
       }
+    } recover {
+      case ex =>
+        log.warn("Unable to connect to database under config: " + configPath, ex)
```
Comment (mcovarr) [inline src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala]: This seems more like an error than a warning

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+  import TableDrivenPropertyChecks._
+
+  behavior of "WdlValue"
+
+  val wdlValueRawStrings = Table(
+    ("wdlValue", "rawString"),
+    (WdlBoolean(false), "false"),
+    (WdlBoolean(true), "true"),
+    (new WdlFile(Paths.get("hello/world/path")), "hello/world/path"),
```
Comment (mcovarr) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: why `new` here?

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+class WdlValueSpec extends FlatSpec with Matchers {
+
+  import TableDrivenPropertyChecks._
+
+  behavior of "WdlValue"
+
+  val wdlValueRawStrings = Table(
+    ("wdlValue", "rawString"),
+    (WdlBoolean(false), "false"),
+    (WdlBoolean(true), "true"),
```
Comment (mcovarr) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: These two lines beget IntelliJ warnings for unnamed boolean parameters, fortunately `WdlBoolean.True` and `WdlBoolean.False` can help with that.  :smile:

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+  val wdlFloatTruncated = Table(
+    ("wdlValue", "rawString", "validateFloat"),
+    (WdlFloat(Double.MinPositiveValue), "4.9E-324", { d: Double => d == 0.0 }),
+    (WdlFloat(Double.NaN), "NaN", { d: Double => d.isNaN }),
+    (WdlFloat(Double.MaxValue), "1.7976931348623157E308", { d: Double => d.isPosInfinity }),
+    (WdlFloat(Double.MinValue), "-1.7976931348623157E308", { d: Double => d.isNegInfinity }))
+
+  forAll(wdlFloatTruncated) { (wdlValue, rawString, validateFloat) =>
+    it should s"truncate a Wdl${wdlValue.wdlType.toWdlString} to/from raw string '$rawString'" in {
```
Comment (mcovarr) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: Not sure I follow why we're testing float truncation, or even how except for the first case?

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+      val toRawString = wdlValue.toRawString
+      toRawString should be(rawString)
+
+      val wdlType = wdlValue.wdlType
+      val fromRawString = wdlType.fromRawString(toRawString)
+      fromRawString shouldNot be(wdlValue)
```
Comment (mcovarr) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: What does this assertion tell us about the correctness of the results?

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+      fromRawString.wdlType should be(wdlType)
+    }
+  }
+
+  val notImplementRawString = Table(
+    "wdlValue",
+    WdlObject(Map("key" -> WdlString("value"))),
+    WdlNamespace.load(SampleWdl.HelloWorld.wdlSource()))
+
+  forAll(notImplementRawString) { wdlValue =>
```
Comment (mcovarr) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: Is there a point to testing the non-implementation of these methods on `WdlObject` and `WdlNamespace`?  Other than boosting coverage?  :smile:

File: src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala [hunk 12]
```diff
@@ -36,6 +35,24 @@ object SlickDataAccess {
   implicit class StringToClob(val str: String) extends AnyVal {
     def toClob: Clob = new SerialClob(str.toCharArray)
   }
+
+  implicit class ConfigWithUniqueSchema(val config: Config) extends AnyVal {
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: I'm behind the times, can you explain what this is doing (and why)?

File: src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala [hunk 13]
```diff
@@ -70,11 +87,18 @@ class SlickDataAccess(databaseConfig: Config, val dataAccess: DataAccessComponen
     //   used in key specification without a key length
     //
     // Perhaps we'll use a more optimized data type for UUID's bytes in the future, as a FK, instead auto-inc cols
+    //
+    // The value `${slick.uniqueSchema}` may used in the url, in combination with `slick.createSchema = true`, to
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: "may be"

Also, I think I now get the "why" in my previous comment, although it'd be good to see it in plain english to make sure I understand

File: src/main/scala/cromwell/server/WorkflowManagerSystem.scala [hunk 14]
```diff
@@ -7,7 +7,6 @@ import cromwell.engine.workflow.WorkflowManagerActor
 trait WorkflowManagerSystem {
   val systemName = "cromwell-system"
   implicit val actorSystem = ActorSystem(systemName)
+  lazy val dataAccess: DataAccess = DataAccess()
```
Comment (geoffjentry) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: lazy because of Config interaction?

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    (WdlExpression.fromString("10 % 3"), "10 % 3"),
+    (WdlExpression.fromString("10-6.7"), "10 - 6.7"),
+    (WdlExpression.fromString(""" 1 + "String" """), """1 + "String""""),
+    (WdlExpression.fromString("a + b"), "a + b"),
+    (WdlExpression.fromString("a(b, c)"), "a(b, c)"),
+    (WdlExpression.fromString("\"a\" + \"b\""), "\"a\" + \"b\""),
+    (WdlExpression.fromString("a.b.c"), "a.b.c"))
+
+  forAll(wdlExpressionRawStrings) { (wdlValue, rawString) =>
+    it should s"resemble a Wdl${wdlValue.wdlType.toWdlString} to/from raw string '$rawString'" in {
```
Comment (mcovarr) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: maybe 
```
trait WdlValue {
...
  def typeName: String = wdlType.getClass().getSimpleName()
...
}
...
  it should s"resemble a ${wdlValue.typeName} to/from blah"
 
```

File: src/main/scala/cromwell/server/CromwellServer.scala [hunk 16]
```diff
@@ -18,7 +17,12 @@ import scala.util.{Failure, Success}
 // Note that as per the language specification, this is instantiated lazily and only used when necessary (i.e. server mode)
 object CromwellServer extends DefaultWorkflowManagerSystem {
   val conf = ConfigFactory.parseFile(new File("/etc/cromwell.conf"))
-  private lazy val realDataAccess = DataAccess()
+
+  // NOTE: Currently the this.dataAccess is passed in to this.workflowManagerActor.
+  // The actor could technically restart with the same instance of the dataAccess,
+  // So, we're not shutting down dataAccess during this.workflowManagerActor.postStop() nor this.service.postStop().
+  // Not sure otherwise when this server is really shutting down, so this.dataAccess currently never explicitly closed.
+  // Shouldn't be an issue unless perhaps test code tries to launch multiple servers and leaves dangling connections.
```
Comment (mcovarr) [inline src/main/scala/cromwell/server/CromwellServer.scala]: Yeah, we have some lifecycle stuff to figure out.  Thanks for documenting what's happening here.  :smile:

File: src/test/scala/cromwell/engine/WorkflowManagerActorSpec.scala (hunk 1 above)
Comment (geoffjentry) [inline src/test/scala/cromwell/engine/WorkflowManagerActorSpec.scala]: :+1: for loan pattern

aside: Beyond thinking [this talk](https://www.parleys.com/tutorial/unsung-heroes-less-fashionable-patterns-scala) from Dick Wall is worthwhile he does have an amusing anecdote about bioinformaticists abusing the loan pattern (just after 13:00 if you want to skip the rest)

File: src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala (hunk 12 above)
Comment (kshakir) [inline src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: This function below (optionally) inserts a unique string into the config value `url`, that is seconds later used internally by slick to connect to the database.

When this code snippet activates for the default in memory database (see also reference.conf) this allows two things:

First, quick google searches weren't clear on a one-liner way to make slick behave like liquibase, only updating a schema. If special checks weren't added, each time the schema creation ran, after the first time, it would bomb out on "table already exists". There were a couple of search hits on how to peek in the database to see if the tables existed, but I ultimately backed off on this because...

Secondly, providing a clean, unique database each call to `DataAccess()`/`new SlickDataAccess()` leaves the behavior `SlickDataAccess` similar to the old `DummyDataAccess`. The coercion bug wasn't discovered earlier as `DummyDataAccess` was being used for tests. Now they can just use `DataAccess()` and a clean database will be created for the test.

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala (hunk 11 above)
Comment (kshakir) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: Sanity check: if the behavior of the main code ever changes (for the better?), this test will catch it.

Comment (mcovarr) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: nm, on further reflection I like this.  :smile:  If these types ever gain real raw string implementations this will force the implementer to address the testing issue.

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala (hunk 10 above)
Comment (kshakir) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: Next push adds a comment to line above:

> // Test that this is a special conversion, and is not
> // expected to be equal after a round-trip conversion.

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala (hunk 9 above)
Comment (kshakir) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: True. Bad name. Renamed to `wdlFloatSpecials`.

File: src/test/scala/cromwell/binding/values/WdlValueSpec.scala (hunk 7 above)
Comment (kshakir) [inline src/test/scala/cromwell/binding/values/WdlValueSpec.scala]: Whoops, didn't notice it was a `case class` and needed the `Path` constructor. Thanks!

File: src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala (hunk 6 above)
Comment (kshakir) [inline src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala]: This led to "red herrings" in the logs for those without mysql setup locally. Ideally we all should for regression testing, but I didn't want it to be mandatory. So technically this is a warning.

File: src/main/scala/cromwell/server/WorkflowManagerSystem.scala (hunk 14 above)
Comment (kshakir) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: Didn't want to start create a database pool until someone needs the actual variable. The primary consumer of this variable, `workflowManagerActor`, is `lazy`. So I made this `lazy` as well.

Comment (geoffjentry) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: :+1:

Comment (geoffjentry) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: Tho might not be a bad idea to add a comment to that effect. I wouldn't say doing so was necessary though, but I'll probably come across and wonder "Why lazy?" again a week from now :)

File: src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala (hunk 12 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: Gotcha. Can you leave some commentary here describing the situation a bit more?

**Comment by kshakir** [comment]
Ready for next round of review. Also, please let me know if I didn't address anything from the first round with either a sufficient fix or a comment.

File: src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala (hunk 6 above)
Comment (mcovarr) [inline src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala]: would it be possible to adjust severity based on config representing MySQL versus HSQLDB?

**Comment by mcovarr** [comment]
With the addition of some comments in the spots where there's discussion, looks good to me. :smile:

File: src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala (hunk 6 above)
Comment (kshakir) [inline src/test/scala/cromwell/engine/db/slick/TestSlickDatabase.scala]: This is the second half of a two part boolean:

```scala
canConnect = testRequired || testDatabase.isValidConnection.futureValue
```

For the hsqldb databases, `testRequired` is `true`, so this short circuit won't activate, and the tests will throw errors.

I updated the method comment. If we would ever like to support users running cromwell on other dbs, we can add those as similar optional tests tests for those database engines too.

**Comment by geoffjentry** [comment]
:+1: from me
